hardware/kubernetes.go

```diff
@@ -0,0 +1,132 @@
+// between the cluster and internal caches. Consumers can wait for the initial sync using WaitForCachesync().
+// See k8s.io/client-go/tools/clientcmd for constructing *rest.Config objects.
+func NewKubernetesClient(config *rest.Config, namespace string) (*KubernetesClient, error) {
+	opts := tink.GetServerOptions()
+	opts.Namespace = namespace
+
+	// Use a manager from the tink project so we can take advantage of the indexes and caching it configures.
+	// Once started, we don't really need any of the manager capabilities hence we don't store it in the
+	// KubernetesClient
+	manager, err := tink.NewManager(config, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	managerCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		if err := manager.Start(managerCtx); err != nil {
+			panic(err)
+		}
+	}()
+
+	client := NewKubernetesClientWithClient(manager.GetClient())
+	client.close = cancel
+	client.waitForCacheSync = manager.GetCache().WaitForCacheSync
+
+	return client, nil
+}
+
+// NewKubernetesClientWithClient creates a new KubernetesClient instance that uses client to find resources. The
+// Close() and WaitForCacheSync() methods of the returned client are noops.
+func NewKubernetesClientWithClient(client ListerClient) *KubernetesClient {
+	return &KubernetesClient{
+		client:           client,
+		close:            func() {},
+		waitForCacheSync: func(context.Context) bool { return true },
+	}
+}
+
+// Close stops synchronization with the cluster. Subsequent calls to all of k's methods have undefined behavior.
+func (k *KubernetesClient) Close() {
+	k.close()
+}
+
+// WaitForCacheSync waits for the internal client cache to synchronize.
+func (k *KubernetesClient) WaitForCacheSync(ctx context.Context) bool {
+	return k.waitForCacheSync(ctx)
+}
+
+// IsHealthy always returns true.
+func (k *KubernetesClient) IsHealthy(context.Context) bool {
+	return true
+}
+
+// ByIP retrieves a hardware resource associated with ip.
+func (k *KubernetesClient) ByIP(ctx context.Context, ip string) (Hardware, error) {
+	var hw tinkv1alpha1.HardwareList
+	err := k.client.List(ctx, &hw, crclient.MatchingFields{
+		tink.HardwareIPAddrIndex: ip,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hw.Items) == 0 {
+		return nil, fmt.Errorf("no hardware with ip '%v'", ip)
+	}
+
+	if len(hw.Items) > 1 {
+		names := make([]string, len(hw.Items))
+		for i, item := range hw.Items {
+			names[i] = item.Name
+		}
+		return nil, fmt.Errorf("multiple hardware with ip '%v': [%v]", ip, strings.Join(names, ", "))
+	}
+
+	return &hardware{hw.Items[0]}, nil
+}
+
+// Watch is unimplemented.
+func (k *KubernetesClient) Watch(context.Context, string) (Watcher, error) {
+	return nil, errors.New("kubernetes client: watch is unimplemented")
+}
+
+type hardware struct {
+	tinkv1alpha1.Hardware
+}
+
+func (h hardware) Export() ([]byte, error) {
+	return json.Marshal(h.Hardware.Spec)
```

Comment:
What structure does Export expect here? Is the CRD spec going to be ingested correctly?

Reply:
[See my other reply](https://github.com/tinkerbell/hegel/pull/82#discussion_r851406714). 

I think there are some inconsistencies. There will likely be a need to translate between the CRD spec and the export expectations but some fields match up already hence I was able to validate things "work".

Reply:
I've encoded the minimal expectations into a new export construct to be safe (there were some problems anyhow so this fixes them). 

It would be useful to move away entirely from this approach but it likely means breaking compatibility with the custom metadata rules that can be added via an env var on startup. I suspect others are using that capability which makes it difficult to detach from the jq parsing approach and head toward something that's more clearly defined.